PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "Defines.hpp"
+
+namespace Poly {
+	template<typename V>
+	class Optional {
+	public:
+		using ValueType = V;
+	public:
+		Optional() : initialized(false), dummy{} {}
+		Optional(const V&  value) : initialized(true), value_storage(value) {}
+		Optional(      V&& value) : initialized(true), value_storage(std::move(value)) {}
+		~Optional() { if (this->HasValue()) { this->Value().~V(); } }
+
+	public:
+		Optional(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				::new(&this->Value()) V(std::move(other.Value()));
+
+				other.Value().~V();
+				other.initialized = false;
+			} else {
+				this->initialized = false;
+			}
+		}
+		Optional& operator=(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				this->Value() = std::move(other.Value());
+
+				other.Value().~V();
+				other.initialized = false;
+			} else if (this->HasValue()) {
+				this->Value().~V();
+				this->initialized = false;
+			}
+			return *this;
+		}
+
+	public:
+		bool HasValue() const { return this->initialized; }
+		operator bool() const { return this->HasValue(); }
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&& Value() &&     { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return std::move(this->value_storage); }
+
+	public:
+		V TakeValue() {
```

Comment:
Steal is even weirder, haha. I mean, the more natural way to say it is "take (the) value out". If that doesn't work there is still "take (the) value for ourselves" ;)